apisix/admin/routes.lua

```diff
@@ -168,143 +162,4 @@ local function check_conf(id, conf, need_id)
-        core.log.error("failed to get route[", key, "] from etcd: ", err)
-        return 503, {error_msg = err}
-    end
-
-    utils.fix_count(res.body, id)
-    return res.status, res.body
-end
-
-
-function _M.post(id, conf, sub_path, args)
-    local id, err = check_conf(id, conf, false)
-    if not id then
-        return 400, err
-    end
-
-    local key = "/routes"
-    utils.inject_timestamp(conf)
-    local res, err = core.etcd.push(key, conf, args.ttl)
-    if not res then
-        core.log.error("failed to post route[", key, "] to etcd: ", err)
-        return 503, {error_msg = err}
-    end
-
-    return res.status, res.body
-end
-
-
-function _M.delete(id)
-    if not id then
-        return 400, {error_msg = "missing route id"}
-    end
-
-    local key = "/routes/" .. id
-    -- core.log.info("key: ", key)
-    local res, err = core.etcd.delete(key)
-    if not res then
-        core.log.error("failed to delete route[", key, "] in etcd: ", err)
-        return 503, {error_msg = err}
-    end
-
-    return res.status, res.body
-end
-
-
-function _M.patch(id, conf, sub_path, args)
-    if not id then
-        return 400, {error_msg = "missing route id"}
-    end
-
-    if conf == nil then
-        return 400, {error_msg = "missing new configuration"}
-    end
-
-    if not sub_path or sub_path == "" then
-        if type(conf) ~= "table"  then
-            return 400, {error_msg = "invalid configuration"}
-        end
-    end
-
-    local key = "/routes"
-    if id then
-        key = key .. "/" .. id
-    end
-
-    local res_old, err = core.etcd.get(key)
-    if not res_old then
-        core.log.error("failed to get route [", key, "] in etcd: ", err)
-        return 503, {error_msg = err}
-    end
-
-    if res_old.status ~= 200 then
-        return res_old.status, res_old.body
-    end
-    core.log.info("key: ", key, " old value: ",
-                  core.json.delay_encode(res_old, true))
-
-    local node_value = res_old.body.node.value
-    local modified_index = res_old.body.node.modifiedIndex
-
-    if sub_path and sub_path ~= "" then
-        local code, err, node_val = core.table.patch(node_value, sub_path, conf)
-        node_value = node_val
-        if code then
-            return code, err
-        end
-        utils.inject_timestamp(node_value, nil, true)
-    else
-        node_value = core.table.merge(node_value, conf)
-        utils.inject_timestamp(node_value, nil, conf)
-    end
-
-    core.log.info("new conf: ", core.json.delay_encode(node_value, true))
-
-    local id, err = check_conf(id, node_value, true)
-    if not id then
-        return 400, err
-    end
-
-    local res, err = core.etcd.atomic_set(key, node_value, args.ttl, modified_index)
-    if not res then
-        core.log.error("failed to set new route[", key, "] to etcd: ", err)
-        return 503, {error_msg = err}
-    end
-
-    return res.status, res.body
-end
-
-
-return _M
+return resource.new("routes", "route", check_conf)
```

Comment:
I think about it a bit.
Maybe it would be better to pass an opt like:
```
{
        name = name,
        kind = kind,
        check_conf = check_conf
    }
```

so people don't need to remember the order.